### PR TITLE
[GL] Algorithm for computing line segment intersections in the xy-plane

### PR DESCRIPTION
**Pull request by TomFischer**
This PR not only contains the algorithm itself, there are also 
- some new point and line segment generators used within the 
- tests of the algorithm.

The lineSegmentIntersection2d algorithm is an important to fix the GeoMapper, see issue #944.

### CODE REVIEW

File: GeoLib/AnalyticalGeometry.cpp [hunk 1]
```diff
@@ -515,4 +528,77 @@ GeoLib::Polygon rotatePolygonToXY(GeoLib::Polygon const& polygon_in,
...
+
+	MathLib::GaussAlgorithm<
+		MathLib::DenseMatrix<double, std::size_t>, std::vector<double>> solver(mat);
+	solver.solve(rhs);
+	if (0 <= rhs[1] && rhs[1] <= 1.0) {
+		return { MathLib::Point3d{std::array<double,3>{{
+				c[0]+rhs[1]*(d[0]-c[0]), c[1]+rhs[1]*(d[1]-c[1]),
+				c[2]+rhs[1]*(d[2]-c[2])}} } };
+	} else {
+		return std::vector<MathLib::Point3d>(); // parameter s not in the valid range
```
Comment (endJunction) [inline GeoLib/AnalyticalGeometry.cpp]: At this point if s in not in the valid range, is it an error or just normal case with empty intersection?
If error, please add a message.

File: GeoLib/AnalyticalGeometry.h [hunk 2]
```diff
@@ -299,6 +299,10 @@ bool parallel(MathLib::Vector3 v, MathLib::Vector3 w);
 bool lineSegmentIntersect (const GeoLib::Point& a, const GeoLib::Point& b,
         const GeoLib::Point& c, const GeoLib::Point& d, GeoLib::Point& s);
 
+std::vector<MathLib::Point3d>
+lineSegmentIntersect2d(MathLib::Point3d const& a, MathLib::Point3d const& b,
```
Comment (endJunction) [inline GeoLib/AnalyticalGeometry.h]: Docu. Especially on return values. Mention also 'ab' is a segment and 'cd' is a segment...

Comment (endJunction) [inline GeoLib/AnalyticalGeometry.h]: Use LineSegment instead of points if possible.

File: Tests/GeoLib/AutoCheckGenerators.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		    std::array<double, 3>{{center[0] + radius * std::cos(angle),
+		                           center[1] + radius * std::sin(angle), 0.0}}};
+	}
+
+	Gen generator;
+	MathLib::Point3d const center;
+	double const radius;
+};
+
+// reflect point p on the point c
```
Comment (endJunction) [inline Tests/GeoLib/AutoCheckGenerators.h]: ... in x-y plane.

File: Tests/GeoLib/AutoCheckGenerators.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "autocheck/autocheck.hpp"
+#include "MathLib/Point3d.h"
+#include "MathLib/Vector3.h"
+#include "GeoLib/LineSegment.h"
+
+namespace autocheck
+{
+
+// generates points on a circle
```
Comment (endJunction) [inline Tests/GeoLib/AutoCheckGenerators.h]: ... in x-y plane.

File: GeoLib/AnalyticalGeometry.cpp (hunk 1 above)
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.cpp]: The line segments aren't intersecting each other. This is a valid result of the algorithm.

File: Tests/GeoLib/TestLineSegmentIntersect2d.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+	MathLib::Vector3 const translation_vector2 = {0,0,0};
+	PairSegmentGenerator pair_segment_generator2 = PairSegmentGenerator{
+	    segment_generator1,
+	    std::bind(ac::translate, translation_vector2, std::placeholders::_1)};
+
+	ac::gtest_reporter gtest_reporter;
+};
+
+TEST_F(LineSegmentIntersect2dTest, RandomSegmentOrientationIntersecting)
```
Comment (endJunction) [inline Tests/GeoLib/TestLineSegmentIntersect2d.cpp]: Please add a short description what setup is tested and what we expect. Not a math proof but one sentence.

File: GeoLib/AnalyticalGeometry.h (hunk 2 above)
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.h]: In principle it is possible to use `LineSegment` objects here. The main reason I want to postpone this is that `GeoLib::Point`s are used for the definition of  a `LineSegment`. I want to use this algorithm also for line segments build from `MeshLib::Node`s. At the moment `LineSegment` is only for testing purposes.

**Comment by endJunction** [comment]
:fast_forward:

**Comment by endJunction** [comment]
Jenkins, test this please.

**Comment by endJunction** [comment]
Jenkins, test this please.

**Comment by bilke** [comment]
:+1:

**Pull request merged**